qiskit/algorithms/fidelities/base_fidelity.py

```diff
@@ -0,0 +1,253 @@
+        else:
+            # ensure 2d list
+            if not isinstance(values, list):
+                values = [values]
+            if len(values) > 0 and not isinstance(values[0], list):
+                values = [values]
+            return values
+
+    def _check_qubits_match(self, circuit_1: QuantumCircuit, circuit_2: QuantumCircuit) -> None:
+        """
+        Checks that the number of qubits of 2 circuits matches.
+        Args:
+            circuit_1: (Parametrized) quantum circuit
+            circuit_2: (Parametrized) quantum circuit
+
+        Raises:
+            ValueError: when ``circuit_1`` and ``circuit_2`` don't have the same number of qubits.
+        """
+
+        if circuit_1 is not None and circuit_2 is not None:
+            if circuit_1.num_qubits != circuit_2.num_qubits:
+                raise ValueError(
+                    f"The number of qubits for the left circuit ({circuit_1.num_qubits}) "
+                    f"and right circuit ({circuit_2.num_qubits}) do not coincide."
+                )
+
+    @abstractmethod
+    def _create_fidelity_circuit(
+        self, circuit_1: QuantumCircuit, circuit_2: QuantumCircuit
+    ) -> QuantumCircuit:
+        """
+        Implementation-dependent method to create a fidelity circuit
+        from 2 circuit inputs.
+        Args:
+            circuit_1: (Parametrized) quantum circuit
+            circuit_2: (Parametrized) quantum circuit
+
+        Returns:
+            The fidelity quantum circuit corresponding to circuit_1 and circuit_2.
+        """
+        raise NotImplementedError
+
+    def _set_circuits(
+        self,
+        circuits_1: Sequence[QuantumCircuit] | None = None,
+        circuits_2: Sequence[QuantumCircuit] | None = None,
+    ) -> None:
+        """
+        Update the list of fidelity circuits to be evaluated.
+        These circuits represent the state overlap between pairs of input circuits,
+        and their construction depends on the fidelity method implementations.
+
+        Args:
+            circuits_1: (Parametrized) quantum circuits.
+            circuits_2: (Parametrized) quantum circuits.
+
+        Raises:
+            ValueError: if the length of the input circuit lists doesn't match.
+        """
+
+        if not len(circuits_1) == len(circuits_2):
+            raise ValueError(
+                f"The length of the first circuit list({len(circuits_1)}) "
+                f"and second circuit list ({len(circuits_2)}) does not coincide."
+            )
+
+        circuits = []
+        for (circuit_1, circuit_2) in zip(circuits_1, circuits_2):
+
+            circuit = self._circuit_cache.get((id(circuit_1), id(circuit_2)))
+
+            if circuit is not None:
+                circuits.append(circuit)
+            else:
+                self._check_qubits_match(circuit_1, circuit_2)
+
+                # re-parametrize input circuits
+                left_parameters = ParameterVector("x", circuit_1.num_parameters)
```

Comment:
We could save some time if we only re-parameterize one circuit, or no circuit if it's not required. But we can also leave that for a follow up

Reply:
I would leave that for a follow up, but good point :) Thanks!